File: object_relations/app/models/customer.rb
class Customer

  @@all = []

  attr_reader :name

  def initialize(name)
    @name = name
    @@all << self
  end

  def add_review(review_text, restaurant)
    Review.new(review_text, restaurant, self)
  end

  def self.find_by_name(name)
    all.find {|restaurant| restaurant.name==name}
  end

end


# Review#customer
# Review#restaurant

# Restaurant#customers
# Restaurant#reviews


# customer = customer.add_review('it was good', Restaurant.new)
# a new review tied ot the restaurant
# and tied to the customer

File: javascript/models/customer.js
const Customer = ( function Customer(){
  var counter = 0

  return class {
    constructor(name) {
      this.id = ++counter
      this.name = name
      store.customers.push(this)
    }

    addReview(text, restaurant){
      return new Review(text, this.id, restaurant.id)
    }

    reviews(){
      return store.reviews.filter((review) => {
        return review.customerId===this.id
      })
    }

    restaurants(){
      return this.reviews().map((review) => {
        return review.restaurant()[0]
      })
    }

    static all() {
      return store.customers
    }
  }
}
())
